### PR TITLE
Fix split_non_commuting bug for Hamiltonians with identity term

### PR DESCRIPTION
**Pull request by austingmhuang**
**Context:**
Described in https://github.com/PennyLaneAI/pennylane/issues/5924, but to summarize, `split_non_commuting` does not handle Identity terms with non-trainable coefficients well, leading to an unexpected `NonDifferentiableError`.

Although the coefficients of Identity terms (also known as `offset`) is assumed to be a float given the type-hinting of `_sum_terms()`, it seems that this is not the case. It is also not the case for `coeffs` which is assumed to be a list of floats but can actually be a list of tensors.

**Description of the Change:**
Fixes https://github.com/PennyLaneAI/pennylane/issues/5924 by setting `requires_grad` to `True` for offsets when autograd is in use. 

**Benefits:**
Fixes https://github.com/PennyLaneAI/pennylane/issues/5924

**Possible Drawbacks:**
May have unintended side effects, or perhaps we would like to coerce `offset` and `coeffs` to be the types in the type hint.

**Related GitHub Issues:**
[sc-67508]

### CODE REVIEW

**Review by soranjh — APPROVED** [review]
Thanks @austingmhuang for fixing this. 

[optional] You might update the type hinting issue in this PR as well.

File: pennylane/transforms/split_non_commuting.py [hunk 1]
```diff
@@ -384,17 +384,17 @@ def _split_ham_with_grouping(tape: qml.tape.QuantumScript):
 
 def _split_using_qwc_grouping(
     tape: qml.tape.QuantumScript,
-    single_term_obs_mps: dict[MeasurementProcess, tuple[list[int], list[float]]],
-    offsets: list[float],
+    single_term_obs_mps: dict[MeasurementProcess, tuple[list[int], list[Union[float, TensorLike]]]],
+    offsets: list[Union[float, TensorLike]],
```
Comment (albi3ro) [inline pennylane/transforms/split_non_commuting.py:388]:
```suggestion
    offsets: list[TensorLike],
```
float is included in the definition of `TensorLike`.

File: pennylane/transforms/split_non_commuting.py [hunk 2]
```diff
@@ -384,17 +384,17 @@ def _split_ham_with_grouping(tape: qml.tape.QuantumScript):
...
 ):
     """Split tapes using group_observables in the Pauli module.
 
     Args:
         tape (~qml.tape.QuantumScript): The tape to be split.
-        single_term_obs_mps (Dict[MeasurementProcess, Tuple[List[int], List[float]]]): A dictionary
+        single_term_obs_mps (Dict[MeasurementProcess, Tuple[List[int], List[Union[float, TensorLike]]]]): A dictionary
```
Comment (albi3ro) [inline pennylane/transforms/split_non_commuting.py:394]:
```suggestion
        single_term_obs_mps (Dict[MeasurementProcess, Tuple[List[int], List[TensorLike]]]): A dictionary
```

**Review by albi3ro — CHANGES_REQUESTED** [review]
Even if this fixes this particular problem, it will probably cause worse problems that will be extremely difficult to track down. Making something trainable can force us to do a lot more work later. Work the user explicitly didn't want.  If someone marked something as trainable, but for some reason it starts being trainable, that could be really difficult to track down.

I'd prefer to find a different solution instead.

**Comment by austingmhuang** [comment]
> Even if this fixes this particular problem, it will probably cause worse problems that will be extremely difficult to track down. Making something trainable can force us to do a lot more work later. Work the user explicitly didn't want. If someone marked something as trainable, but for some reason it starts being trainable, that could be really difficult to track down.
> 
> I'd prefer to find a different solution instead.

I'm not sure what the best fix for this particular problem is actually. 

How about this? We could add an additional check to see if the coeffs are trainable. Since the offset is just the coeff for the identity term, we can assume that the differentiability of the offset based on the differentiability of the coeffs. If coeff is trainable -> set `requires_grad` to `True` -- otherwise we set it to `False`.

**Comment by astralcai** [comment]
> Even if this fixes this particular problem, it will probably cause worse problems that will be extremely difficult to track down. Making something trainable can force us to do a lot more work later. Work the user explicitly didn't want. If someone marked something as trainable, but for some reason it starts being trainable, that could be really difficult to track down.
> 
> I'd prefer to find a different solution instead.

Unfortunately, after some investigation, it turned out that the reason why [this solution](https://github.com/PennyLaneAI/pennylane/pull/5945) worked was that it also caused things to become trainable, specifically via [this line](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/numpy/tensor.py#L170).

**Comment by austingmhuang** [comment]
> Unfortunately, after some investigation, it turned out that the reason why [this solution](https://github.com/PennyLaneAI/pennylane/pull/5945) worked was that it also caused things to become trainable, specifically via [this line](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/numpy/tensor.py#L170).

Is there a way to fix these two bugs without causing things to become trainable?

**Comment by soranjh** [comment]
> Since the offset is just the coeff for the identity term, we can assume the differentiability of the offset based on the differentiability of the coeffs. If coeff is trainable -> set requires_grad to True -- otherwise we set it to False.

This seems like a reasonable solution.

**Comment by astralcai** [comment]
I think there's a deeper issue here. The error is always raised from [here](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/numpy/tensor.py#L308), which forbids turning an untrainable tensor to an `ArrayBox`. When the code executes to [this line](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/transforms/split_non_commuting.py#L719), since there are other things in the circuit that are trainable, `sum_dot_products` is an `ArrayBox`, so executing `+` takes us [here](https://github.com/HIPS/autograd/blob/master/autograd/tracer.py#L31). This is a recursive function that takes values out of an `ArrayBox` [here](https://github.com/HIPS/autograd/blob/master/autograd/tracer.py#L37), performs the underlying operation [here](https://github.com/HIPS/autograd/blob/master/autograd/tracer.py#L44), and repackages the result into an `ArrayBox` [here](https://github.com/HIPS/autograd/blob/master/autograd/tracer.py#L46).

The issue is, when performing the underlying operation, we have on one hand the value that was taken out of the `ArrayBox` (which is now just a regular value), and on the other hand a tensor that was explicitly marked untrainable, the result is also an untrainable tensor. When autograd tries to repackage this result into an `ArrayBox`, it dispatches to [this function](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/numpy/tensor.py#L290), which now complains that it doesn't allow an untrainable tensor to be converted into an `ArrayBox`.

**Comment by astralcai** [comment]
> > Since the offset is just the coeff for the identity term, we can assume the differentiability of the offset based on the differentiability of the coeffs. If coeff is trainable -> set requires_grad to True -- otherwise we set it to False.
> 
> This seems like a reasonable solution to me. cc @albi3ro @astralcai

Maybe instead of assuming that all coeffs have the same trainability, we look at whether `summed_dot_products` is trainable, so something like:

```python
...
def _sum_terms(res: ResultBatch, coeffs: list[float], offset: float, shape: tuple) -> Result:
    ...
    summed_dot_products = qml.math.sum(qml.math.stack(dot_products), axis=0)
    if qml.math.get_interface(offset) == "autograd" and qml.math.requires_grad(summed_dot_products):
        offset = qml.math.array(offset)
    return summed_dot_products + offset
```

**Pull request merged**